src/PusherCrypto.php

```diff
@@ -0,0 +1,145 @@
+<?php
+
+namespace Pusher;
+
+class PusherCrypto
+{
+    private $encryption_key = '';
+
+    // The prefix any e2e channel must have
+    const ENCRYPTED_PREFIX = 'private-encrypted-';
+
+    // The exact length the user specified key must be
+    const SECRET_KEY_LENGTH = 32;
+
+    /**
+     * Checks if a given channel is an encrypted channel.
+     *
+     * @param string $channel the name of the channel
+     *
+     * @return bool true if channel is an encrypted channel
+     */
+    public static function is_encrypted_channel($channel)
+    {
+        return substr($channel, 0, strlen(self::ENCRYPTED_PREFIX)) === self::ENCRYPTED_PREFIX;
+    }
+
+    /**
+     * Initialises a PusherCrypto instance.
+     *
+     * @param string $encryption_key the SECRET_KEY_LENGTH key that will be used for key derivation.
+     */
+    public function __construct($encryption_key)
+    {
+        if (function_exists('sodium_crypto_secretbox')) {
+            if (strlen($encryption_key) === SODIUM_CRYPTO_SECRETBOX_KEYBYTES) {
+                $this->encryption_key = $encryption_key;
+
+                return;
+            } else {
+                throw new PusherException('Your end to end encryption key must be 32 chars long');
+            }
+        }
+
+        throw new PusherException('To use end to end encryption, you must either be using PHP 7.2 or greater or have installed the libsodium-php extension for php < 7.2.');
+    }
+
+    /**
+     * Decrypts a given event.
+     *
+     * @param object $event an object that has an encrypted data property and a channel property.
+     *
+     * @return object the event with a decrypted payload, or false if decryption was unsuccessful.
+     */
+    public function decrypt_event($event)
+    {
+        $encrypted_payload = explode(':', $event->data);
+        $nonce = base64_decode($encrypted_payload[1]);
+        $payload = base64_decode($encrypted_payload[2]);
+        $shared_secret = $this->generate_shared_secret($event->channel, $this->encryption_key);
+        $decrypted_payload = $this->decrypt_payload($payload, $nonce, $shared_secret);
+        if (!$decrypted_payload) {
+            return false;
+        }
+        $event->data = $decrypted_payload;
+
+        return $event;
+    }
+
+    /**
+     * Derives a shared secret from the secret key and the channel to broadcast to.
+     *
+     * @param string $channel the name of the channel
+     *
+     * @return string a SHA256 hash (encoded as base64) of the channel name appended to the encryption key
+     */
+    public function generate_shared_secret($channel)
+    {
+        if ($channel == '') {
+            return false;
+        }
```

Comment:
Exception would be better yes but isn't returning false just a PHP convention?

Reply:
Unfortunately I don't think there is much consistency in the PHP ecosystem. I think the "old way" was to return false, and the "new way" is to throw exceptions. I'd rather stick with the convention used in the rest of the library, but unfortunately it looks like we use a mixture of the two: e.g. [exception](https://github.com/pusher/pusher-http-php/blob/master/src/Pusher.php#L309-L311) and [returning false](https://github.com/pusher/pusher-http-php/blob/master/src/Pusher.php#L615).

IMO you should use an exception here since it conveys the cause of the error to the user. It looks like we only return `false` when an HTTP response code != 200.

Reply:
Agreed!